### PR TITLE
Add new configuration item `listing_table_ignore_subdirectory`

### LINKED ISSUE #8524: `select count(*) from '*.parquet'` scans all parquet file recursively including subdirectory
### Describe the bug

I find `select count(*) from '*.parquet'` not only scan the parquet file in current directory, but it also recursively scan all the parquet file in subdirectory. I wonder is this behavior by design or a bug.

### To Reproduce

I tried with three parquet file, and two of them are in subdir.
```
$$ tree
.
├── subdir
│   ├── file1.parquet
│   └── file2.parquet
└── users.parquet

2 directories, 3 files
```

`users.parquet` has 2 record, `file1.parquet` has 1 record and `file2.parquet` has 1 record.
`select count(*) from '*.parquet'` get 4
```
$$ datafusion-cli -c "select count(*) from '*.parquet'"
DataFusion CLI v33.0.0
+----------+
| COUNT(*) |
+----------+
| 4        |
+----------+
1 row in set. Query took 0.058 seconds.

$$ datafusion-cli -c "select count(*) from 'users.parquet'"
DataFusion CLI v33.0.0
+----------+
| COUNT(*) |
+----------+
| 2        |
+----------+
1 row in set. Query took 0.002 seconds.

$$ datafusion-cli -c "select count(*) from 'subdir/*.parquet'"
DataFusion CLI v33.0.0
+----------+
| COUNT(*) |
+----------+
| 2        |
+----------+
1 row in set. Query took 0.002 seconds.
```

### Expected behavior

I try same query in duckdb, and duckdb only scan parquet file in current directory
```
$$ duckdb -c "select count(*) from '*.parquet'"
┌──────────────┐
│ count_star() │
│    int64     │
├──────────────┤
│            2 │
└──────────────┘
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Asura7969**
## Which issue does this PR close?

Closes #8524.

## Rationale for this change

Consistent behavior with duckdb and hive

## What changes are included in this PR?

add config `listing_table_ignore_subdirectory`
When scanning file paths, whether to ignore subdirectory files, ignored by default (true)

NOTE: When reading a partitioned table, `listing_table_ignore_subdirectory` is always equal to false, even if set to true
## Are these changes tested?

add sqllogictest

## Are there any user-facing changes?

`ListingTableUrl::contains` method add parameters

### CODE REVIEW

File: datafusion/core/src/execution/context/parquet.rs [hunk 1]
```diff
@@ -109,7 +109,7 @@ mod tests {
             .read_parquet(
                 // it was reported that when a path contains // (two consecutive separator) no files were found
                 // in this test, regardless of parquet_test_data() value, our path now contains a //
-                format!("{}/..//*/alltypes_plain*.parquet", parquet_test_data()),
```
Comment (Asura7969) [inline datafusion/core/src/execution/context/parquet.rs:112]: Should we add regex matching to directories?🤔

Comment (alamb) [reply datafusion/core/src/execution/context/parquet.rs:112]: I don't fully understand this question

File: datafusion/core/src/datasource/listing/url.rs [hunk 2]
```diff
@@ -424,6 +433,13 @@ mod tests {
         let b = ListingTableUrl::parse("../bar/./foo/../baz").unwrap();
         assert_eq!(a, b);
         assert!(a.prefix.as_ref().ends_with("bar/baz"));
+
+        let url = ListingTableUrl::parse("../foo/*.parquet").unwrap();
```
Comment (alamb) [inline datafusion/core/src/datasource/listing/url.rs:437]: I am probably missing something here, but how does this test the new code? I don't see it passing in `ignore_subdirectory`

File: datafusion/sqllogictest/test_files/information_schema.slt [hunk 3]
```diff
@@ -150,6 +150,7 @@ datafusion.execution.aggregate.scalar_update_factor 10
 datafusion.execution.batch_size 8192
 datafusion.execution.coalesce_batches true
 datafusion.execution.collect_statistics false
+datafusion.execution.ignore_subdirectory true
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/information_schema.slt:153]: Could we use a name that gives some context about when the `ignore_subdirectory` is actually used?

For example, maybe like `listing_table_ignore_subdirectory`. Or maybe even it is time to create a whole new category of configuration for listing tables `datafusion.listing_table.ignore_subdirectory` 🤔

**Review by alamb — COMMENTED** [review]
Thank you @Asura7969  -- this code looks good to me, but I think we need some test coverage (I am not sure the test in this PR covers the changes)

Could you maybe create a sqllogictest test for this? So like write three files:

Like this 
```
table/1.parquet
table/subdir/2.parquet
table/3.parquet
```
And then verify that with `ignore_subdirectory` to true (default) only the rows from `1.parquet` and `3.parquet` appear in the output but  when the setting is set to false then the rows from all three files are present?

There is an example of creating and reading parquet files in a sqllogictest in these two files:

https://github.com/apache/arrow-datafusion/blob/main/datafusion/sqllogictest/test_files/parquet.slt (👋  @hiltontj  -- we are already using it!)

https://github.com/apache/arrow-datafusion/blob/main/datafusion/sqllogictest/test_files/repartition_scan.slt

File: datafusion/core/src/datasource/listing/url.rs (hunk 2 above)
Comment (Asura7969) [reply datafusion/core/src/datasource/listing/url.rs:437]: Yes, it's really not obvious here (actually in `ListingTableUrl.contains`), I would create a sqllogictest as you suggested

File: datafusion/sqllogictest/test_files/information_schema.slt (hunk 3 above)
Comment (Asura7969) [reply datafusion/sqllogictest/test_files/information_schema.slt]: I used `listing_table_ignore_subdirectory`

File: datafusion/sqllogictest/test_files/csv_files.slt [hunk 4]
```diff
@@ -63,3 +63,60 @@ id6 value"6
...
+--TableScan: partition_csv_table projection=[]
+physical_plan
+AggregateExec: mode=Final, gby=[], aggr=[COUNT(*)]
+--CoalescePartitionsExec
+----AggregateExec: mode=Partial, gby=[], aggr=[COUNT(*)]
+------RepartitionExec: partitioning=RoundRobinBatch(4), input_partitions=2
+--------CsvExec: file_groups={2 groups: [[WORKSPACE_ROOT/datafusion/core/tests/data/partitioned_table/c_date=2018-11-13/timestamps.csv], [WORKSPACE_ROOT/datafusion/core/tests/data/partitioned_table/c_date=2018-12-13/timestamps.csv]]}, has_header=false
+
+query I
+select count(*) from partition_csv_table;
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/csv_files.slt:117]: I don't understand what this test is testing -- in both cases the table has 4 rows (aka there is no data in a subdirectory to ignore, right)?

File: datafusion/sqllogictest/test_files/parquet.slt [hunk 5]
```diff
@@ -276,6 +276,118 @@ LIMIT 10;
...
+------ParquetExec: file_groups={2 groups: [[WORKSPACE_ROOT/datafusion/sqllogictest/test_files/scratch/parquet/test_table/0.parquet, WORKSPACE_ROOT/datafusion/sqllogictest/test_files/scratch/parquet/test_table/1.parquet], [WORKSPACE_ROOT/datafusion/sqllogictest/test_files/scratch/parquet/test_table/2.parquet, WORKSPACE_ROOT/datafusion/sqllogictest/test_files/scratch/parquet/test_table/subdir/3.parquet]]}
+
+# scan file: 0.parquet 1.parquet 2.parquet 3.parquet
+query I
+select count(*) from t1_with_subdirectory;
+----
+12
+
+query I
+select count(*) from t2_with_subdirectory;
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/parquet.slt:387]: It is cool to see the different rows but I don't understand the need for all the different tables and explain plans

I think we can get coverage by simply creating the equivalent of `t2_with_subdirectory` and showing that it returns 12 rows when

```sql
set datafusion.execution.listing_table_ignore_subdirectory = false;
```

And 9 when 

```sql
set datafusion.execution.listing_table_ignore_subdirectory = true;
```

File: datafusion/sqllogictest/test_files/csv_files.slt (hunk 4 above)
Comment (Asura7969) [reply datafusion/sqllogictest/test_files/csv_files.slt:117]: When reading a partitioned table, `listing_table_ignore_subdirectory` is always equal to false, even if the default is true

File: datafusion/sqllogictest/test_files/csv_files.slt [hunk 6]
```diff
@@ -63,3 +63,60 @@ id6 value"6
...
+----
+4
+
+statement ok
+DROP TABLE partition_csv_table
+
+statement ok
+set datafusion.execution.listing_table_ignore_subdirectory = true;
```
Comment (Asura7969) [inline datafusion/sqllogictest/test_files/csv_files.slt:91]: When reading a partitioned table, listing_table_ignore_subdirectory is always equal to false, even if set to true

File: datafusion/sqllogictest/test_files/parquet.slt (hunk 5 above)
Comment (Asura7969) [reply datafusion/sqllogictest/test_files/parquet.slt:387]: Thank you for your suggestion, I understand

File: datafusion/common/src/config.rs [hunk 7]
```diff
@@ -273,6 +273,11 @@ config_namespace! {
         /// memory consumption
         pub max_buffered_batches_per_output_file: usize, default = 2
 
+        /// When scanning file paths, whether to ignore subdirectory files,
+        /// ignored by default (true), when reading a partitioned table,
+        /// `listing_table_ignore_subdirectory` is always equal to false, even if set to true
```
Comment (Asura7969) [inline datafusion/common/src/config.rs:278]: I updated the description information, do you agree? @alamb

File: datafusion/sqllogictest/test_files/parquet.slt [hunk 8]
```diff
@@ -276,6 +276,39 @@ LIMIT 10;
 0 2014-08-27T14:00:00Z Timestamp(Millisecond, Some("UTC"))
 0 2014-08-27T14:00:00Z Timestamp(Millisecond, Some("UTC"))
 
+# Test config listing_table_ignore_subdirectory:
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/parquet.slt:279]: 👍

**Review by alamb — APPROVED** [review]
Thank you @Asura7969  -- this looks great

File: datafusion/common/src/config.rs (hunk 7 above)
Comment (alamb) [reply datafusion/common/src/config.rs:278]: I don't understand what this is trying to say 🤔

Comment (Asura7969) [reply datafusion/common/src/config.rs:278]: image
```sql
## read partition table

CREATE EXTERNAL TABLE csv_with_timestamps (
  name VARCHAR,
  ts TIMESTAMP,
  c_date DATE,
)
STORED AS CSV
PARTITIONED BY (c_date)
LOCATION '../core/tests/data/partitioned_table';

set datafusion.execution.listing_table_ignore_subdirectory = true;

select count(*) from partition_tbale;    ## return 4
```

Comment (alamb) [reply datafusion/common/src/config.rs:278]: Ah, got it -- thank you -- I will propose a clarification in a follow on PR

**Pull request merged**